src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java

```diff
@@ -0,0 +1,369 @@
+
+    runWithTrampoline(v -> {
+      ProtonReceiver receiver = connection.get().createReceiver(address, opts);
+
+      if (receiverOptions != null) {
+        if (receiverOptions.getQos() != null) {
+          receiver.setQoS(ProtonQoS.valueOf(receiverOptions.getQos().toUpperCase()));
+        }
+
+        configureTheSource(recOpts, receiver);
+      }
+
+      new AmqpReceiverImpl(address, this, recOpts, receiver, handler, completionHandler);
+    });
+    return this;
+  }
+
+  private void configureTheSource(AmqpReceiverOptions receiverOptions, ProtonReceiver receiver) {
+    org.apache.qpid.proton.amqp.messaging.Source source = (org.apache.qpid.proton.amqp.messaging.Source) receiver
+      .getSource();
+
+    List<String> capabilities = receiverOptions.getCapabilities();
+    if (!capabilities.isEmpty()) {
+      source.setCapabilities(capabilities.stream().map(Symbol::valueOf).toArray(Symbol[]::new));
+    }
+
+    if (receiverOptions.isDurable()) {
+      source.setExpiryPolicy(TerminusExpiryPolicy.NEVER);
+      source.setDurable(TerminusDurability.UNSETTLED_STATE);
+    }
+  }
+
+  @Override
+  public AmqpConnection createSender(String address, Handler<AsyncResult<AmqpSender>> completionHandler) {
+    Objects.requireNonNull(address, "The address must be set");
+    return createSender(address, new AmqpSenderOptions(), completionHandler);
+  }
+
+  @Override
+  public AmqpConnection createSender(String address, AmqpSenderOptions options,
+    Handler<AsyncResult<AmqpSender>> completionHandler) {
+    if (address == null && !options.isDynamic()) {
+      throw new IllegalArgumentException("Address must be set if the link is not dynamic");
+    }
+
+    Objects.requireNonNull(completionHandler, "The completion handler must be set");
+    runWithTrampoline(x -> {
+
+      ProtonSender sender;
+      if (options != null) {
+        ProtonLinkOptions opts = new ProtonLinkOptions();
+        opts.setLinkName(options.getLinkName());
+        opts.setDynamic(options.isDynamic());
+
+        sender = connection.get().createSender(address, opts);
+        sender.setAutoDrained(options.isAutoDrained());
+      } else {
+        sender = connection.get().createSender(address);
+      }
+
+      // TODO durable?
+
+      AmqpSenderImpl.create(sender, this, completionHandler);
+    });
+    return this;
+  }
+
+  @Override
+  public AmqpConnection createAnonymousSender(Handler<AsyncResult<AmqpSender>> completionHandler) {
+    Objects.requireNonNull(completionHandler, "The completion handler must be set");
+    runWithTrampoline(x -> {
+      ProtonSender sender = connection.get().createSender(null);
+      AmqpSenderImpl.create(sender, this, completionHandler);
+    });
+    return this;
+  }
+
+  @Override
+  public AmqpConnection closeHandler(Handler<AmqpConnection> remoteCloseHandler) {
+    this.connection.get().closeHandler(pc -> {
+      if (remoteCloseHandler != null) {
+        runWithTrampoline(x -> remoteCloseHandler.handle(this));
```

Comment:
As this replaces the underlying connections close handler directly, it should also run the onDisconnect bits like the default handler set while this object is created does. Thats need so that if this were a remotely-initiated close, rather than response to a locally-initiated close, then our side actually sends its close frame back (the protocol is symmetric).